sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteGroupingAnalyticsWithoutKeys.scala

```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeReference, AttributeSet, Literal, NamedExpression, VirtualColumn}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * In grouping analytics (e.g., `GROUPING SETS`) with an empty grouping set, we need to follow
+ * the non-key aggregate semantics; it generates a single row having an initial aggregated value
+ * (e.g., 0 for `COUNT`) from an empty input.
+ * For example, a query below must return a row (NULL, NULL, 0) from
+ * an empty input (`empty_input`):
+ *
+ * {{{
+ *   SELECT k1, k2, COUNT(v)
+ *     FROM empty_input t(k1, k2, v)
+ *     GROUP BY GROUPING SETS ((), (k1, k2))
+ * }}}
+ *
+ * To comply with the semantics, this rule rewrites the following (pseudo) resolved logical plan
+ * as an union query of aggregates with/without keys as follows:
+ *
+ * {{{
+ *   Aggregate(
+ *     groupingExprs = ['k1, 'k2, 'spark_grouping_id],
+ *     aggregateExprs = ['k1, 'k2, COUNT('v)],
+ *     output = ['k1, 'k2, 'cnt],
+ *     child = Expand(
+ *       projections = [('k1, 'k2, 'v, 0), (null, null, 'v, 3)],
+ *       output = ['k1, 'k2, 'v, 'spark_grouping_id]),
+ *       child = LocalTableScan [...]))
+ * }}}
+ *
+ * It is transformed into an union plan below:
+ *
+ * {{{
+ *   Union([
+ *     Aggregate(
+ *       groupingExprs = ['k1, 'k2, 'spark_grouping_id],
+ *       aggregateExprs = ['k1, 'k2, COUNT('v)],
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Expand(
+ *         projections = [('k1, 'k2, 'v, 0)],
+ *         output = ['k1, 'k2, 'v, 'spark_grouping_id]),
+ *         child = LocalTableScan [...])),
+ *     Aggregate(
+ *       groupingExprs = [],
+ *       aggregateExprs = [null, null, COUNT('v)],
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Project(
+ *         projectList = ['v],
+ *         child = LocalTableScan [...]))]
+ * }}}
+ */
+object RewriteGroupingAnalyticsWithoutKeys extends Rule[LogicalPlan] {
+
+  private val virtualGroupingColumnNames =
+    Seq(VirtualColumn.groupingIdName, VirtualColumn.groupingPosName)
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+    case agg @ Aggregate(groupingExprs, aggExprs, expand @ Expand(projs, childOutput, child)) =>
+      val groupByAttrMap = AttributeSet.fromAttributeSets(groupingExprs.map {
+        case ar: AttributeReference if virtualGroupingColumnNames.contains(ar.name) =>
+          AttributeSet.empty
+        case e =>
+          e.references
+      })
+      val (groupsWithoutKeys, groupsWithKeys) = {
+        val groupByIndices = childOutput.zipWithIndex.filter { case (a, _) =>
+          groupByAttrMap.contains(a)
+        }
+        projs.partition { groupingSet =>
```

Comment:
each projection doesn't only include grouping expressions. It is not a set too. Maybe just name it as `proj`.

Reply:
Yea, it looks ok.